fix(ivy): remove DOM nodes from their real parent vs saved parent

jelbourn opened the pull request:
Currently, DOM node removal called `removeChild` on the saved parent
node when destroying a component. However, this will fail if the
component has been manually moved in the DOM. This change makes the
removal always use the node's real `parentNode` and ignore the provided
`parent`.

Reference: FW-977

kara commented on packages/core/src/render3/node_manipulation.ts:
```diff
@@ -589,8 +589,13 @@ function nativeAppendOrInsertBefore(
-  isProceduralRenderer(renderer) ? renderer.removeChild(parent as RElement, child, isHostElement) :
-                                   parent.removeChild(child);
+  if (isProceduralRenderer(renderer)) {
+    renderer.removeChild(parent as RElement, child, isHostElement);
```
Could we not pass in `child.parentNode` here and avoid changing implementation of the dom renderer?

JoostK replied:
There's a helper fn `nativeParentNode` that seems appropriate to use here. That would also correspond with ViewEngine logic:

https://github.com/angular/angular/blob/885f1af509eb7d9ee049349a2fe5565282fbfefb/packages/core/src/view/util.ts#L268-L270

kara replied:
Yeah, seems like we could do this as well, if that's what you mean:

```ts
renderer.removeChild(renderer.parentNode(child), child, isHostElement);
```

kara commented on packages/core/test/render3/view_container_ref_spec.ts:
```diff
@@ -2087,4 +2087,64 @@ describe('ViewContainerRef', () => {
+        const componentRef =
+            this.viewContainerRef.createComponent(componentFactory, this.viewContainerRef.length);
+
+        // Manually nove the loaded component to some arbitrary DOM node.
```
```suggestion
        // Manually move the loaded component to some arbitrary DOM node.
```

pkozlowski-opensource commented on packages/core/test/render3/view_container_ref_spec.ts:
```diff
@@ -2089,6 +2088,66 @@ describe('ViewContainerRef', () => {
 
   });
 
+  describe('with manually moved nodes', () => {
```
I would rather write this test as a `TestBed` test as it would be less brittle.

jelbourn replied:
I wasn't aware that was even an option. Just using TedBed in this same spec?

kara replied:
It would probably be cleaner to add it to an existing TestBed spec file. I think in this case `view_injector_integration_spec.ts` would be the best one.

pkozlowski-opensource commented on packages/core/src/render3/node_manipulation.ts:
```diff
@@ -598,8 +598,16 @@ function nativeAppendOrInsertBefore(
     renderer: Renderer3, parent: RElement, child: RNode, isHostElement?: boolean): void {
-  isProceduralRenderer(renderer) ? renderer.removeChild(parent as RElement, child, isHostElement) :
-                                   parent.removeChild(child);
+  if (isProceduralRenderer(renderer)) {
```
If we do this logic here in nativeRemoveChild it will mean that we still execute all the logic to find a logical render parent and this is useless. I would rather modify the `removeChild` function and there lookup a parent from the DOM, instead of executing `const parentNative = getRenderParent(childTNode, currentView);`

Then the `null` check be in there as well and you wouldn't need to modify renderer

jelbourn replied:
This is what I originally had and @kara asked to change it. @kara what are your thoughts here?

kara replied:
@jelbourn I believe @pkozlowski-opensource is actually talking about our version of `removeChild` here:
https://github.com/angular/angular/blob/master/packages/core/src/render3/node_manipulation.ts#L696

I'd agree with him that we should update that function to remove the logic that finds the render parent.

jelbourn replied:
I totally misread Pawel's original comment; I hadn't seen that `getRenderParent` function earlier. 

In making the change to avoid calling that, I didn't want to move the "get parent node" bit into `removeChild` because then `nativeRemoveChild` would have the assumption that it's the right parent node baked into it. So, I instead changed `removeChild` to `removeNode` and made it just a pass-through to `nativeRemoveChild`, which also now does not have `parent` in its arguments. 

Let me know what you think.

pkozlowski-opensource reviewed (changes requested):
I think that we should rather:
* modify the `removeChild` function (see my in-line comments);
* have a `TestBed` test on top of / instead of renderer3 runtime test

Happy to help if needed but I would like to very much avoid the situation where we do _both_ logical render parent lookup _and_ lookup in the DOM.

jelbourn committed:
fix(ivy): remove DOM nodes from their real parent vs saved parent

Currently, DOM node removal called `removeChild` on the saved parent
node when destroying a component. However, this will fail if the
component has been manually moved in the DOM. This change makes the
removal always use the node's real `parentNode` and ignore the provided
`parent`.

jelbourn reviewed (commented):
Comments addressed, let me know what you think

pkozlowski-opensource committed:
refactor(ivy): move around logic of getting render parent

pkozlowski-opensource commented on packages/core/src/render3/node_manipulation.ts:
```diff
@@ -593,13 +593,19 @@ function nativeAppendOrInsertBefore(
+      renderer.removeChild(renderParent, child, isHostElement);
+    }
+  } else {
+    // We intentionally don't use the given parent node since it may no longer
```
So the idea behind the `native[...]()` functions in this file that those are here to basically abstract a renderer type being used. So far those functions didn't have any logic and I would love to keep it this way.

I would rather prefer that we keep `nativeRemoveChild` "dumb" and move all the "logic" into the `removeChild` function. I find this approach easier to follow and IMO the code is shorter / simpler (less checks / branching logic).

I've pushed a fixup commit https://github.com/angular/angular/pull/28455/commits/4d7d430efb7e939ca16331145406e5add999ea36 to your  branch with the changes I'm describing above, please review.

pkozlowski-opensource commented on packages/core/src/render3/node_manipulation.ts:
```diff
@@ -692,14 +698,9 @@ export function getBeforeNodeForView(index: number, views: LView[], containerNat
-  if (parentNative) {
-    nativeRemoveChild(currentView[RENDERER], parentNative, childEl);
-  }
+export function removeNode(childTNode: TNode, childEl: RNode, currentView: LView): void {
```
Actually, I've pushed a fixup commit https://github.com/angular/angular/pull/28455/commits/4d7d430efb7e939ca16331145406e5add999ea36 to your  branch with the changes I'm describing above, please review.

kara reviewed (approved):
LGTM

pkozlowski-opensource reviewed (approved):
LGTM with the fixup commit but we can iterate more if @jelbourn has objections

jelbourn commented:
I'm good with the proposed changes

matsko commented:
http://test/OCL:232530052:BASE:232530123:1549396551445:4021aaca
